src/java/org/apache/cassandra/index/sai/disk/vector/VectorMemtableIndex.java

```diff
@@ -164,64 +160,56 @@ else if (primaryKey.compareTo(maximumKey) > 0)
 
-            Set<PrimaryKey> resultKeys = isMaxToken ? primaryKeys.tailSet(left, leftInclusive) : primaryKeys.subSet(left, leftInclusive, right, rightInclusive);
+        var pkIterator = searchInternal(context, qv, keyRange, graph.size(), threshold,
+                                        (keys) -> this.filterPrimaryKeys(qv, threshold, keys),
+                                        this::nodeScoreToPrimaryKey);
 
-            if (resultKeys.isEmpty())
-                return RangeIterator.empty();
-
-            int bruteForceRows = maxBruteForceRows(limit, resultKeys.size(), graph.size());
-            logger.trace("Search range covers {} rows; max brute force rows is {} for memtable index with {} nodes, LIMIT {}",
-                         resultKeys.size(), bruteForceRows, graph.size(), limit);
-            Tracing.trace("Search range covers {} rows; max brute force rows is {} for memtable index with {} nodes, LIMIT {}",
-                          resultKeys.size(), bruteForceRows, graph.size(), limit);
-            if (resultKeys.size() <= bruteForceRows)
-                return new ReorderingRangeIterator(new PriorityQueue<>(resultKeys));
-            else
-                bits = new KeyRangeFilteringBits(keyRange);
-        }
+        PriorityQueue<PrimaryKey> keyQueue = new PriorityQueue<>();
+        while (pkIterator.hasNext())
+            keyQueue.add(pkIterator.next());
 
-        var keyQueue = graph.search(qv, limit, expr.getEuclideanSearchThreshold(), bits);
         if (keyQueue.isEmpty())
             return RangeIterator.empty();
         return new ReorderingRangeIterator(keyQueue);
     }
 
     @Override
-    public OrderIterator orderBy(QueryContext queryContext, Expression expr, AbstractBounds<PartitionPosition> keyRange, int limit)
+    public ScoredPrimaryKeyIterator orderBy(QueryContext context, Expression expr, AbstractBounds<PartitionPosition> keyRange, int limit)
     {
         assert expr.getOp() == Expression.Op.ANN : "Only ANN is supported for vector search, received " + expr.getOp();
 
+        float[] qv = expr.lower.value.vector;
+
+        var result = searchInternal(context, qv, keyRange, limit, 0,
+                                    (keys) -> orderPrimaryKeys(qv, keys),
+                                    this::nodeScoreToScoredPrimaryKey);
+        return new WrappedScoredPrimaryKeyIterator(result);
+    }
+
+    private Iterator<PrimaryKey> nodeScoreToPrimaryKey(SearchResult.NodeScore nodeScore)
+    {
+        return graph.keysFromOrdinal(nodeScore.node).stream().iterator();
+    }
+
+    private Iterator<ScoredPrimaryKey> nodeScoreToScoredPrimaryKey(SearchResult.NodeScore nodeScore)
+    {
+        return graph.keysFromOrdinal(nodeScore.node).stream().map(k -> new ScoredPrimaryKey(k, nodeScore.score)).iterator();
+    }
+
+    private <T> Iterator<T> searchInternal(QueryContext context, float[] queryVector, AbstractBounds<PartitionPosition> keyRange, int limit, float threshold,
```

Comment:
nit: let's either include all parameters on one line, or switch to one parameter per line, instead of a mix

Reply:
(same comment for `search`)

Reply:
I think I did this in several places. I'll fix them all.